File: battle/data/CustomUnit.java
package common.battle.data;

import common.io.json.JsonClass;
import common.io.json.JsonField;
import common.pack.Identifier;
import common.util.Data;
import common.util.pack.Soul;
import common.util.unit.Form;
import org.jcodec.common.tools.MathUtil;

import java.util.ArrayList;

@JsonClass
public class CustomUnit extends CustomEntity implements MaskUnit, Cloneable {

	public Form pack;

	@JsonField
	public int price, resp, back, front, limit;

	@JsonField(gen = JsonField.GenType.GEN)
	public PCoin pcoin = null;

	public CustomUnit() {
		rep = new AtkDataModel(this);
		atks = new AtkDataModel[1];
		atks[0] = new AtkDataModel(this);
		width = 320;
		speed = 8;
		hp = 1000;
		hb = 1;
		traits = new ArrayList<>();
		price = 50;
		resp = 60;
		back = 0;
		front = 9;
		death = new Identifier<>(Identifier.DEF, Soul.class, 0);
	}

	@Override
	public int getBack() {
		return back;
	}

	@Override
	public int getFront() {
		return front;
	}

	@Override
	public Orb getOrb() {
		return pack.orbs;
	}

	@Override
	public Form getPack() {
		return pack;
	}

	@Override
	public int getPrice() {
		return price;
	}

	@Override
	public int getRespawn() {
		return resp;
	}

	@Override
	public int getLimit() {
		return limit;
	}

	@Override
	public PCoin getPCoin() { return pcoin; }

	@Override
	public void importData(MaskEntity de) {
		super.importData(de);

		if (de instanceof MaskUnit) {
			MaskUnit mu = (MaskUnit) de;

			price = mu.getPrice();
			resp = mu.getRespawn();
			back = Math.min(mu.getBack(), mu.getFront());
			front = Math.max(mu.getBack(), mu.getFront());
			limit = mu.getLimit();

			PCoin p = mu.getPCoin();
			if (p == null)
				return;

			ArrayList<int[]> info = p.info;
			pcoin = new PCoin(this);
			pcoin.max = p.max;
			for (int i = 0; i < info.size(); i++) {
				int[] j = info.get(i).clone();
				int[] data = Data.PC_CORRES[j[0]];
				j[1] = pcoin.max[i] = data[3] > -1 ? 1 : Math.max(1, j[1]);
				j[13] = MathUtil.clip(j[13], 0, 1);
				if (data[3] > -1) {
					j[0] = data[3];
					j[2] = j[3] = 100;
				}
				if (j[0] == 65 || j[0] == 56) {
					int[] lvl = new int[] { j[4], j[5] };
					int[] max = new int[] { j[6], j[7] };
					int[] min = new int[] { j[8], j[9] };
					j[4] = max[0] / 4;
					j[5] = max[1] / 4;
					j[6] = (max[0] + min[0]) / 4;
					j[7] = (max[1] + min[1]) / 4;
					j[8] = lvl[0];
					j[9] = lvl[1];
				}
				pcoin.info.add(j);
			}
		}
	}

	@Override
	public CustomUnit clone() {
		CustomUnit ans = (CustomUnit) Data.err(super::clone);
		ans.importData(this);
		ans.pack = getPack();
		ans.getPack().anim = getPack().anim;
		return ans;
	}
}

File: battle/data/CustomUnit.kt
package common.battle.data

import common.io.InStream
import common.io.assets.Admin
import common.io.assets.Admin.StaticPermitted
import common.io.assets.AssetLoader
import common.io.assets.AssetLoader.AssetHeader
import common.io.assets.AssetLoader.AssetHeader.AssetEntry
import common.io.json.JsonClass
import common.io.json.JsonEncoder
import common.io.json.JsonField
import common.io.json.Test
import common.io.json.Test.JsonTest_0.JsonD
import common.io.json.Test.JsonTest_2
import common.pack.Source.AnimLoader
import common.pack.Source.ResourceLocation
import common.pack.Source.SourceAnimLoader
import common.pack.Source.SourceAnimSaver
import common.pack.Source.Workspace
import common.pack.Source.ZipSource
import common.util.Data
import common.util.stage.EStage
import common.util.stage.StageMap
import common.util.stage.StageMap.StageMapInfo
import common.util.unit.Form
import common.util.unit.UnitLevel
import io.BCPlayer
import page.JL
import page.anim.AnimBox
import page.support.ListJtfPolicy
import page.support.SortTable
import page.view.ViewBox
import page.view.ViewBox.Conf
import page.view.ViewBox.Controller
import page.view.ViewBox.VBExporter

@JsonClass
class CustomUnit : CustomEntity(), MaskUnit {
    var pack: Form? = null

    @JsonField
    var price: Int

    @JsonField
    var resp: Int
    fun fillData(ver: Int, `is`: InStream) {
        zread(ver, `is`)
    }

    override fun getBack(): Int {
        return 9
    }

    override fun getFront(): Int {
        return 0
    }

    override fun getOrb(): Orb? {
        return pack!!.orbs
    }

    override fun getPack(): Form? {
        return pack
    }

    override fun getPrice(): Int {
        return price
    }

    override fun getRespawn(): Int {
        return resp
    }

    override fun importData(de: MaskEntity) {
        super.importData(de)
        if (de is MaskUnit) {
            val mu: MaskUnit = de
            price = mu.getPrice()
            resp = mu.getRespawn()
        }
    }

    private fun zread(`val`: Int, `is`: InStream) {
        var `val` = `val`
        `val` = Data.Companion.getVer(`is`.nextString())
        if (`val` >= 400) `zread$000400`(`is`)
    }

    private fun `zread$000400`(`is`: InStream) {
        zreada(`is`)
        price = `is`.nextInt()
        resp = `is`.nextInt()
    }

    init {
        rep = AtkDataModel(this)
        atks = arrayOfNulls<AtkDataModel>(1)
        atks.get(0) = AtkDataModel(this)
        width = 320
        speed = 8
        hp = 1000
        hb = 1
        type = 0
        price = 50
        resp = 60
    }
}
